src/triage/component/catwalk/predictors.py

```diff
@@ -87,6 +88,34 @@ def _existing_predictions(self, Prediction_obj, session, model_id, matrix_store)
             .filter(Prediction_obj.as_of_date.in_(self._as_of_dates(matrix_store)))
         )
 
+    @db_retry
+    def needs_predictions(self, matrix_store, model_id):
+        """Returns whether or not the given matrix and model are lacking any predictions
+
+        Args:
+            matrix_store (triage.component.catwalk.storage.MatrixStore) A matrix with metadata
+            model_id (int) A database ID of a model
+
+        The way we check is by grabbing all the distinct as-of-dates in the predictions table
+        for this model and matrix. If there are more as-of-dates defined in the matrix's metadata
+        than are in the table, we need predictions
+        """
+        if not self.save_predictions:
+            return False
+        session = self.sessionmaker()
+        prediction_obj = matrix_store.matrix_type.prediction_obj
+        as_of_dates_in_db = set(
+            obj.as_of_date.date()
+            for obj in session.query(prediction_obj).filter_by(
+                model_id=model_id,
+                matrix_uuid=matrix_store.uuid
+            ).distinct(prediction_obj.as_of_date).all()
+        )
+        as_of_dates_needed = set(as_of_date for as_of_date in self._as_of_dates(matrix_store))
+        needed = bool(as_of_dates_needed - as_of_dates_in_db)
```

Comment:
And to bring this bikeshedding to a close :man_shrugging: this could be, (tersely for brevity of the comment):

```suggestion
        needed = bool(
            set(self._as_of_dates(matrix_store))
            .difference(
                as_of_date.date()
                for as_of_date in session.query(…).filter_by(
                    …
                ).distinct(…).value(…)
            )
        )
```

Reply:
I prefer the named variables for clarity here.